specs/agents/tracing-instrumentation-http.md

```diff
@@ -61,6 +61,23 @@ All errors that are captured during a request to an ignored URL are still sent t
 | Dynamic        | `true` |
 | Central config | `true` |
 
+### `transaction_ignore_user_agents` configuration
+
+Used to restrict requests made by certain User-Agents from being instrumented.
+
+This property should be set to a list containing one or more strings.
+When an incoming HTTP request is detected, the `User-Agent` request headers will be tested against each element in this list and if a match is found, no trace will be captured for this request.
+
+NOTE: 
+All errors that are captured during a request made by an ignored user agent are still sent to the APM Server regardless of this setting.
```

Comment:
What is the motivation behind this exception?

Reply:
I guess the motivation is that the fact that you don't want to clutter your traces with such that are not very interesting to you doesn't mean you don't want to see errors related to them. 

This is how it is already implemented today both in the [Node agent](https://www.elastic.co/guide/en/apm/agent/nodejs/current/configuration.html#ignore-user-agents) and in the [Java agent](https://www.elastic.co/guide/en/apm/agent/java/current/config-http.html#config-ignore-user-agents) and it is aligned with the spec for the [`transaction_ignore_urls` config](https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-http.md#transaction_ignore_urls-configuration) that serves pretty much the same purpose.

Reply:
I'm not sure if that's actually what the Java agent would do, though. If there's an exception in a servlet and the request is excluded via `transaction_ignore_urls` or `transaction_ingore_user_agents`, we would not capture the error.

I'm not entirely sure on this but IIRC, in Node.js and Python/Django there's a separate error handler. Errors caught by that handler are always reported, no matter if there's a corresponding transaction. Hence, it's more natural for them to not apply the `transaction_ignore*` settings again in that error handler.

Reply:
The reason I asked is because I thought about it I could not find a real world situation when I would want to see the errors but not any other data that would help me find the root cause for those errors. IMHO finding and discussing those use cases will help better understand configuration options like `transaction_ingore_user_agents`.

Reply:
I'll be damned!
Well, as long as it is in the spec, we will need to remedy.

> Hence, it's more natural for them to not apply the transaction_ignore* settings again in that error handler.

Probably, but it shouldn't be a big issue for us to support that as well. Our internal API already allows capturing errors without a parent, so IIANM, all we need to do is capture the exception in the method exit before aborting.

Reply:
> The reason I asked is because I thought about it I could not find a real world situation when I would want to see the errors but not any other data that would help me find the root cause for those errors.

Assume you have a ping endpoint - it's occurring very frequently and is almost always very quick and it is responsible for 95% of the traffic your server is getting. You may want to ignore that as it may skew your entire service health indicators. However, you still want to know when there are related errors. Makes sense?

Reply:
>  However, you still want to know when there are related errors. Makes sense?

Wouldn't you want to find the root cause for those errors?

Reply:
> You may want to ignore that as it may skew your entire service health indicators. However, you still want to know when there are related errors.

It sounds like you do want service health indicators to take into account failed pings - it's only the successful pings that should be ignored.

Reply:
> Wouldn't you want to find the root cause for those errors?

Yes, but with head-based sampling there may be unreasonable overhead to support that. With tail-based sampling these configs may be less important, but there is still the issue of affecting entire service statistics.

> It sounds like you do want service health indicators to take into account failed pings - it's only the successful pings that should be ignored.

I actually meant to overall service performance, as in:
![Screen Shot 2021-02-03 at 9 50 35](https://user-images.githubusercontent.com/41850454/106715061-544af400-6605-11eb-9f3b-4c94a79c1af7.png)
You may get these not properly reflecting your service due to a single (or some) transaction type. I **assume** this is what these settings are aimed to solve, but I actually don't recall a real use case.
You are right in that we won't get error rate for the ping service, but on the other hand we may have 0.2% error rate if we include it where we would have wanted to see the 20% error rate of the endpoints we care about. 

Bottom line: I don't have a strong opinion either way. Input from others may be useful, but most importantly - let's make a decision and move forward.

Reply:
> I actually meant to overall service performance

Yes, I understood you correctly but thank you for providing more background :)

The issue that I see with separating error events from health indicator is that you essentially split health indicator into two pieces - now just seeing good health indicator (based on transaction events) is not enough - you need to go and also look at Errors tab.

> Assume you have a ping endpoint - it's occurring very frequently and is almost always very quick and it is responsible for 95% of the traffic your server is getting.

I am not sure I agree that it's normal situation when synthetic traffic reaches such volume. IMHO synthetic traffic volume for active end user experience monitoring should be very small compared to the traffic volume from the real users otherwise monitoring itself would cause issues that would have not existed without such monitoring.

Reply:
@SergeyKleyman one possible usage for this configuration would be to ignore known user agents (like bots) that are hitting applications publicly exposed, for example creating lots of noise with 404 errors trying to detect `*.php` endpoints on a Java application without much performance impact. There are probably better ways to deal with those, but that would provide a quite good work-around and keep the noise level low.

Reply:
@SylvainJuge I'm not sure I follow your explanation - would you still want to see those 404 errors at Errors tab? If so could you please explain why?

Reply:
In that particular case, we can assume that end user would like to "hide" those 404 errors, as they aren't really related to any application feature, and we can't control the client side. Being able to filter those enables to focus on the "real" application errors.

Reply:
> In that particular case, we can assume that end user would like to "hide" those 404 errors

So if I understood you correctly your use case would be served better if `transaction_ingore_user_agents` forces the agent NOT to send both transactions and errors, right?
Do you think there are use cases where agent should still send errors even for requests matched by `transaction_ingore_user_agents`?

Reply:
Yes, in this case ignoring the transaction would also mean ignoring the errors, which we don't report outside of transactions/spans (at least in the java agent).